src/main/host/process.rs

```diff
@@ -484,20 +500,236 @@ impl Process {
+    ///
+    /// The pointer is invalidated when one of the Process memory flush methods is called.
+    unsafe fn writable_ptr(
+        process: &Process,
+        ptr: TypedPluginPtr<u8>,
+    ) -> Result<*mut c_void, Errno> {
+        let manager = RefMut::map(process.memory_manager.borrow_mut(), |mm| {
+            mm.as_mut().unwrap()
+        });
+        // SAFETY: We ensure that the `memory` is dropped before the `manager`,
+        // and `Process` ensures that this whole object is dropped before
+        // `MemoryManager` can be moved, freed, etc.
+        let mut manager = unsafe {
+            std::mem::transmute::<RefMut<'_, MemoryManager>, RefMut<'static, MemoryManager>>(
+                manager,
+            )
+        };
+        let memory = manager.memory_ref_mut_uninit(ptr)?;
+        let mut memory = unsafe {
+            std::mem::transmute::<ProcessMemoryRefMut<'_, u8>, ProcessMemoryRefMut<'static, u8>>(
+                memory,
+            )
+        };
+        let vptr = memory.as_mut_ptr() as *mut c_void;
+        let prev = process.unsafe_borrow_mut.borrow_mut().replace(Self {
+            _manager: manager,
+            memory: Some(memory),
+        });
+        assert!(prev.is_none());
+        Ok(vptr)
+    }
+
+    /// Creates a raw mutable pointer, and saves an instance of `Self` into
+    /// `process` for later clean-up.
+    ///
+    /// # Safety
+    ///
+    /// The pointer is invalidated when one of the Process memory flush methods is called.
+    unsafe fn mutable_ptr(
+        process: &Process,
+        ptr: TypedPluginPtr<u8>,
+    ) -> Result<*mut c_void, Errno> {
+        let manager = RefMut::map(process.memory_manager.borrow_mut(), |mm| {
+            mm.as_mut().unwrap()
+        });
+        // SAFETY: We ensure that the `memory` is dropped before the `manager`,
+        // and `Process` ensures that this whole object is dropped before
+        // `MemoryManager` can be moved, freed, etc.
+        let mut manager = unsafe {
+            std::mem::transmute::<RefMut<'_, MemoryManager>, RefMut<'static, MemoryManager>>(
+                manager,
+            )
+        };
+        let memory = manager.memory_ref_mut(ptr)?;
+        let mut memory = unsafe {
+            std::mem::transmute::<ProcessMemoryRefMut<'_, u8>, ProcessMemoryRefMut<'static, u8>>(
+                memory,
+            )
+        };
+        let vptr = memory.as_mut_ptr() as *mut c_void;
+        let prev = process.unsafe_borrow_mut.borrow_mut().replace(Self {
+            _manager: manager,
+            memory: Some(memory),
+        });
+        assert!(prev.is_none());
+        Ok(vptr)
+    }
+
+    /// Free this reference, writing back to process memory.
+    fn flush(&mut self) -> Result<(), Errno> {
+        self.memory.take().unwrap().flush()
+    }
+
+    /// Free this reference without writing back to process memory.
+    fn noflush(&mut self) {
+        self.memory.take().unwrap().noflush()
     }
```

Comment:
Is there a reason these don't take `self`? I think this would allow you to get rid of the `Option`? But if there's a reason for the `Option`, this could use a comment explaining why.

Reply:
Yup, taking `self` is better; done. Still need the `Option`, though.